### ISSUE TITLE
json-serializer normalizeUsingDeclaredMapping overwrites previous attribute if key is not in map

### ISSUE THREAD

**Issue by jpaas** [emberjs/data#3955]
Have a look here... https://github.com/emberjs/data/blob/e1007db415041ef6a9831b731c9f458d11d2ae4f/addon/serializers/json-serializer.js#L735

```js
  normalizeUsingDeclaredMapping: function(modelClass, hash) {
    var attrs = get(this, 'attrs');
    var payloadKey, normalizedKey, key;

    if (attrs) {
      for (key in attrs) {
        payloadKey = this._getMappedKey(key, modelClass);

        if (!hash.hasOwnProperty(payloadKey)) { continue; }

        if (get(modelClass, 'attributes').has(key)) {
          normalizedKey = this.keyForAttribute(key);
        }

        if (get(modelClass, 'relationshipsByName').has(key)) {
          normalizedKey = this.keyForRelationship(key);
        }

        if (payloadKey !== normalizedKey) {
          hash[normalizedKey] = hash[payloadKey];
          delete hash[payloadKey];
        }
      }
    }
  },
```

If I have a payload such as 
```json
{
  "data": {
    "attributes": {
      "in_my_map": "correct",
      "not_in_my_map": "incorrect"
    }
  }
}
```

My model with end up with a value of "incorrect" on the inMyMap property. Why? Because the normalizedKey is not initialized each time through the loop, so it ends up inside the payloadKey != normalizedKey block at the end of the function.

**Comment by jpaas** [comment]
Submitted pull request #3956

**Comment by igorT** [comment]
Is there a use case for having it in your serializer definition if it's not in the model definition. Seems like a case for an assert?

**Comment by igorT** [comment]
cc @wecc

**Comment by jpaas** [comment]
@igorT: In our case we have multiple apps that share the same APIs. One is an administration app that needs all of the attributes of the models in order to edit them. But some of the other apps only care about a few of the attributes and have their own lighter Ember models.

**Comment by wecc** [comment]
@igorT this is definitely a bug that we should fix, see #3956

**Comment by bmac** [comment]
This issue should be fixed by 72a6290b50a5e8c01034d283ce177098ee4ceb29
